### PR TITLE
Support service generation on action folders

### LINKED ISSUE #11: Allow generating messages services in an `action` folder
#### Feature description

`rosidl_typesupport_connext_c` and `rosidl_typesupport_connext_cpp` need to allow generating messages and services in an `action` folder and namespace

#### Implementation considerations

The service templates would need to change to allow service definitions to live in the .actions folder

https://github.com/ros2/rosidl_typesupport_connext/blob/e9d8da960abc691b7f304fabd256d97d006281ee/rosidl_typesupport_connext_c/resource/srv__type_support_c.cpp.em#L47-L53

https://github.com/ros2/rosidl_typesupport_connext/blob/e9d8da960abc691b7f304fabd256d97d006281ee/rosidl_typesupport_connext_cpp/resource/srv__type_support.cpp.em#L39-L43

connects to ros2/rosidl#301

### PR DESCRIPTION
**Pull request by hidmic**
Closes #11. Connected to ros2/rosidl#301. This pull request is part of a series of pull requests that make the necessary changes to allow us to build .srv files outside an `srv` folder.

connects to ros2/rosidl#301

### CODE REVIEW

File: rosidl_typesupport_connext_c/resource/srv__type_support_c.cpp.em
Comment (jacobperron) [inline rosidl_typesupport_connext_c/resource/srv__type_support_c.cpp.em:83]: https://github.com/ros2/rosidl_typesupport_connext/blob/6a6be4317b9cdd3256229d1f7b23030e7b0e71c7/rosidl_typesupport_connext_c/resource/srv__type_support_c.cpp.em#L71-L74

"srv" -> "@(subfolder)" ?

Comment (hidmic) [reply rosidl_typesupport_connext_c/resource/srv__type_support_c.cpp.em:83]: Addressed by https://github.com/ros2/rosidl_typesupport_connext/pull/12/commits/1f857a55ffc68d80b7635154ab844530eda6b7ac.

**Review by jacobperron — APPROVED** [review]
Approving pending CI.

**Review by sloretz — DISMISSED** [review]
I'm unable to get a service server or client to communicate with each other using `RMW_IMPLEMENTATION=rmw_connext_cpp` working with this PR using a service in `test_msgs/action`. I see the client is waiting for the service server to appear indefinitely. When using a service not in an `action` folder they connect fine.

Requesting changes while investigating.

**Comment by hidmic** [comment]
@sloretz How's your setup? How are you reproducing this? I'm unable to (i.e. they connect just fine).

**Comment by sloretz** [comment]
> How's your setup? How are you reproducing this?

@hidmic 

Repos
```
repositories:
  ros2/rcl_interfaces:
    type: git
    url: https://github.com/ros2/rcl_interfaces.git
    version: sloretz/msg_srv_in_action
  ros2/rosidl:
    type: git
    url: https://github.com/ros2/rosidl.git
    version: hidmic/prepare_for_action_generation
  ros2/rosidl_dds:
    type: git
    url: https://github.com/ros2/rosidl_dds.git
    version: hidmic/prepare_for_action_generation
  ros2/rosidl_python:
    type: git
    url: https://github.com/ros2/rosidl_python.git
    version: apojomovsky/allow_generating_msg_and_srv_in_action
  ros2/rosidl_typesupport:
    type: git
    url: https://github.com/ros2/rosidl_typesupport.git
    version: hidmic/prepare_for_action_generation
  ros2/rosidl_typesupport_connext:
    type: git
    url: https://github.com/ros2/rosidl_typesupport_connext.git
    version: hidmic/subfolder_aware_srv_generation
  ros2/rosidl_typesupport_fastrtps:
    type: git
    url: https://github.com/ros2/rosidl_typesupport_fastrtps.git
    version: apojomovsky/subfolder_aware_srv_generation
  ros2/rosidl_typesupport_opensplice:
    type: git
    url: https://github.com/ros2/rosidl_typesupport_opensplice.git
    version: sservulo/prepare_for_action_generation
  ros2/system_tests:
    type: git
    url: https://github.com/ros2/system_tests.git
    version: hidmic/namespace-messages-with-subfolder
```

Client
```python3
from test_msgs.action import Primitives

import rclpy


def main(args=None):
    rclpy.init(args=args)
    node = rclpy.create_node('minimal_client')
    cli = node.create_client(Primitives, 'add_two_ints')

    req = Primitives.Request()
    req.int8_value = 41
    req.int16_value = 1
    while not cli.wait_for_service(timeout_sec=1.0):
        node.get_logger().info('service not available, waiting again...')

    future = cli.call_async(req)
    rclpy.spin_until_future_complete(node, future)

    if future.result() is not None:
        node.get_logger().info(
            'Result of add_two_ints: for %d + %d = %d' %
            (req.int8_value, req.int16_value, future.result().int32_value))
    else:
        node.get_logger().info('Service call failed %r' % (future.exception(),))

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
```

server
```python3
from test_msgs.action import Primitives

import rclpy

g_node = None


def add_two_ints_callback(request, response):
    global g_node
    response.int32_value = request.int8_value + request.int16_value
    g_node.get_logger().info(
        'Incoming request\na: %d b: %d' % (request.int8_value, request.int16_value))

    return response


def main(args=None):
    global g_node
    rclpy.init(args=args)

    g_node = rclpy.create_node('minimal_service')

    srv = g_node.create_service(Primitives, 'add_two_ints', add_two_ints_callback)
    while rclpy.ok():
        rclpy.spin_once(g_node)

    # Destroy the service attached to the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    g_node.destroy_service(srv)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
```

Terminal 1
```
export RMW_IMPLEMENTATION=rmw_connext_cpp
python3 path/to/service_action.py
```

Terminal 2
```
export RMW_IMPLEMENTATION=rmw_connext_cpp
python3 path/to/client_action.py
```

Terminal  1 output is blank.
Terminal 2 output prints that the service is unavailable

```
python3 src/ros2/examples/rclpy/services/minimal_client/client_action.py 
RTI Data Distribution Service Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Co
mmercial User license@rti.com For non-production use only.
Expires on 00-jan-00 See www.rti.com for more information.
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
[INFO] [minimal_client]: service not available, waiting again...
```

Using `rmw_fastrtps_cpp` or `rmw_opensplice_cpp` the service server and client connect fine.

**Comment by hidmic** [comment]
It works just fine on this end. I'll nuke my source build and start fresh.

**Comment by sloretz** [comment]
@hidmic I can't reproduce the issue anymore, and nothing stands out as a possible cause in the PR. I don't know what the cause is, but it doesn't need to hold up this PR.

**Comment by apojomovsky** [comment]
Ok, given that we have green light I'm proceeding to merge this and its related PRs

**Pull request merged**
